src/main/java/org/prebid/server/bidder/huaweiads/HuaweiUtilRequest.java

```diff
@@ -0,0 +1,474 @@
+    }
+
+    private static HuaweiApp resolveHuaweiApp(BidRequest bidRequest) {
+        final App appFromBidReq = bidRequest.getApp();
+
+        if (appFromBidReq != null) {
+            return HuaweiApp.builder()
+                    .version(StringUtils.isBlank(appFromBidReq.getVer()) ? null : appFromBidReq.getVer())
+                    .name(StringUtils.isBlank(appFromBidReq.getName()) ? null : appFromBidReq.getName())
+                    .pkgname(resolvePkgName(appFromBidReq))
+                    .lang(resolveLang(appFromBidReq))
+                    .country(resolveCountry(bidRequest))
+                    .build();
+        }
+        return null;
+    }
+
+    private static HuaweiDevice resolveHuaweiDevice(BidRequest bidRequest, JacksonMapper mapper) {
+        final Device device = bidRequest.getDevice();
+        if (device != null) {
+
+            final ExtUserDataDeviceIdHuaweiAds data = resolveDeviceData(bidRequest, mapper);
+            final boolean dataNotNull = data != null;
+
+            final String oaid = dataNotNull ? resolveId(data.getOaid()) : null;
+            final String gaid = dataNotNull ? resolveId(data.getGaid()) : null;
+            final String imei = dataNotNull ? resolveId(data.getImei()) : null;
+            final String clientTime = dataNotNull ? resolveClientTime(data.getClientTime()) : null;
+
+            if (oaid == null && gaid == null && imei == null) {
+                throw new PreBidException("resolveHuaweiDevice: Imei, Oaid, Gaid are all empty or null");
+            }
+
+            final Integer dnt = device.getDnt();
+            final String country = resolveCountry(bidRequest);
+
+            return HuaweiDevice.builder()
+                    .type(device.getDevicetype())
+                    .useragent(device.getUa())
+                    .os(device.getOs())
+                    .version(device.getOsv())
+                    .maker(device.getMake())
+                    .model(StringUtils.isBlank(device.getModel())
+                            ? DEFAULT_MODEL_NAME
+                            : device.getModel())
+                    .height(device.getH())
+                    .width(device.getW())
+                    .language(device.getLanguage())
+                    .pxratio(device.getPxratio())
+                    .belongCountry(country)
+                    .localeCountry(country)
+                    .ip(device.getIp())
+                    .gaid(gaid)
+                    .oaid(oaid)
+                    .imei(imei)
+                    .clientTime(clientTime)
+                    .isTrackingEnabled(resolveIsTrackingEnabled(dnt, oaid))
+                    .isGaidTrackingEnabled(resolveIsTrackingEnabled(dnt, gaid)).build();
+        }
+        return HuaweiDevice.builder().build();
+    }
+
+    private static String resolveId(List<String> listId) {
+        if (listId == null) {
+            return null;
+        }
+        return !listId.isEmpty() ? listId.get(0) : null;
```

Comment:
can be simplified to return `CollectionUtils.isNotEmpty(listId) ? listId.get(0) : null;`
consider renaming **listId** to smth like **ids**